File: apps/web/src/local/modelInference.ts
import type { ModelConnection } from './modelConnections';
export interface ModelInferenceRecord {
  schema_version: 'local-model-inference-record/v1'; request_id: string; model_id: string; fingerprint: string;
  status: string; started_at: string; finished_at: string; expires_at: string;
  inference_verified: boolean; business_data_sent: false; service_session_only: true;
}
export const inferenceLabels: Record<string, string> = {
  running: '正在等待模型回答…', passed: '模型回答测试通过', response_mismatch: '模型已响应，但未通过测试内容或模型身份核对',
  auth_failed: '认证未通过，请在原框架核对凭据', rate_limited: '服务限流，本次未自动重试',
  unsupported_response: '服务响应与当前测试协议不兼容', service_error: '服务返回异常，本次未自动重试',
  uncertain: '未取得最终结果，服务可能已接收请求；本次未自动重试', configuration_changed: '配置或凭据已变化，本次结果已失效',
};
export function isModelInferenceRecord(value: unknown, item: ModelConnection): value is ModelInferenceRecord {
  if (!value || typeof value !== 'object') return false;
  const v = value as ModelInferenceRecord;
  return v.schema_version === 'local-model-inference-record/v1' && /^mt-[0-9a-f]{32}$/.test(v.request_id)
    && v.model_id === item.id && /^[0-9a-f]{64}$/.test(v.fingerprint) && Object.hasOwn(inferenceLabels, v.status)
    && Number.isFinite(Date.parse(v.started_at)) && v.inference_verified === (v.status === 'passed')
    && v.business_data_sent === false && v.service_session_only === true
    && (v.status === 'running' ? v.finished_at === '' && v.expires_at === ''
      : Number.isFinite(Date.parse(v.finished_at)) && Date.parse(v.expires_at) > Date.parse(v.finished_at));
}

File: apps/agentshield/internal/server/model_inference.go
package server

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"regexp"
	"time"

	"siq-agent-security/apps/agentshield/internal/modelconfig"
)

var inferenceRequestID = regexp.MustCompile(`^mt-[0-9a-f]{32}$`)
var inferenceModelID = regexp.MustCompile(`^mc-[0-9a-f]{32}$`)
var inferenceFingerprint = regexp.MustCompile(`^[0-9a-f]{64}$`)

type modelInferenceRecord struct {
	Schema       string `json:"schema_version"`
	RequestID    string `json:"request_id"`
	ModelID      string `json:"model_id"`
	Fingerprint  string `json:"fingerprint"`
	Status       string `json:"status"`
	StartedAt    string `json:"started_at"`
	FinishedAt   string `json:"finished_at"`
	ExpiresAt    string `json:"expires_at"`
	Inference    bool   `json:"inference_verified"`
	BusinessData bool   `json:"business_data_sent"`
	SessionOnly  bool   `json:"service_session_only"`
}

func (s *Server) findInferenceTarget(id, fp string) (modelconfig.Target, bool) {
	targets, _ := s.modelTargets()
	for _, t := range targets {
		if t.Item.ID == id && t.Item.Fingerprint == fp && t.Item.CanCheck {
			return t, true
		}
	}
	return modelconfig.Target{}, false
}
func (s *Server) modelInferenceTests(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Cache-Control", "no-store")
	if r.Method == http.MethodGet {
		q := r.URL.Query()
		if len(q) != 1 || len(q["model_id"]) != 1 || !inferenceModelID.MatchString(q.Get("model_id")) {
			writeJSON(w, 400, map[string]string{"error": "model_test_request_invalid"})
			return
		}
		s.modelTestMu.Lock()
		var record *modelInferenceRecord
		if id := s.modelTestLatest[q.Get("model_id")]; id != "" {
			copy := s.modelTests[id]
			record = &copy
		}
		s.modelTestMu.Unlock()
		writeJSON(w, 200, map[string]any{"schema_version": "local-model-inference-latest/v1", "record": record})
		return
	}
	if r.Method != http.MethodPost {
		w.WriteHeader(405)
		return
	}
	var body struct {
		Schema      string `json:"schema_version"`
		RequestID   string `json:"request_id"`
		ModelID     string `json:"model_id"`
		Fingerprint string `json:"fingerprint"`
		Confirm     bool   `json:"confirm_test"`
	}
	raw, err := io.ReadAll(http.MaxBytesReader(w, r.Body, 4096))
	if err != nil || r.URL.RawQuery != "" || !exactJSONObject(raw, "schema_version", "request_id", "model_id", "fingerprint", "confirm_test") || json.Unmarshal(raw, &body) != nil || body.Schema != "local-model-inference-create/v1" || !body.Confirm || !inferenceRequestID.MatchString(body.RequestID) || !inferenceModelID.MatchString(body.ModelID) || !inferenceFingerprint.MatchString(body.Fingerprint) {
		writeJSON(w, 400, map[string]string{"error": "model_test_request_invalid"})
		return
	}
	s.modelTestMu.Lock()
	if old, exists := s.modelTests[body.RequestID]; exists {
		s.modelTestMu.Unlock()
		if old.ModelID != body.ModelID || old.Fingerprint != body.Fingerprint {
			writeJSON(w, 409, map[string]string{"error": "model_test_request_changed"})
			return
		}
		writeJSON(w, 200, old)
		return
	}
	if s.modelTestRunning || len(s.modelTests) >= 1024 {
		s.modelTestMu.Unlock()
		writeJSON(w, 409, map[string]string{"error": "model_test_busy_or_limit"})
		return
	}
	if _, ok := s.findInferenceTarget(body.ModelID, body.Fingerprint); !ok {
		s.modelTestMu.Unlock()
		writeJSON(w, 409, map[string]string{"error": "model_configuration_changed"})
		return
	}
	if s.modelTests == nil {
		s.modelTests = map[string]modelInferenceRecord{}
		s.modelTestLatest = map[string]string{}
	}
	record := modelInferenceRecord{Schema: "local-model-inference-record/v1", RequestID: body.RequestID, ModelID: body.ModelID, Fingerprint: body.Fingerprint, Status: "running", StartedAt: time.Now().UTC().Format(time.RFC3339Nano), SessionOnly: true}
	s.modelTests[body.RequestID] = record
	s.modelTestLatest[body.ModelID] = body.RequestID
	s.modelTestRunning = true
	s.modelTestMu.Unlock()
	// Independent of the browser connection. No retries after an uncertain send.
	go func(record modelInferenceRecord) {
		status := "configuration_changed"
		if target, ok := s.findInferenceTarget(record.ModelID, record.Fingerprint); ok {
			status = modelconfig.Infer(context.Background(), target)
			if _, ok = s.findInferenceTarget(record.ModelID, record.Fingerprint); !ok {
				status = "configuration_changed"
			}
		}
		now := time.Now().UTC()
		s.modelTestMu.Lock()
		defer s.modelTestMu.Unlock()
		record.Status = status
		record.Inference = status == "passed"
		record.FinishedAt = now.Format(time.RFC3339Nano)
		record.ExpiresAt = now.Add(5 * time.Minute).Format(time.RFC3339Nano)
		s.modelTests[record.RequestID] = record
		s.modelTestRunning = false
	}(record)
	writeJSON(w, 202, record)
}
